File: 2601-prime-subtraction-operation/2601-prime-subtraction-operation.py
class Solution:
    def primeSubOperation(self, nums: List[int]) -> bool:
        maxn = max(nums)
        prime = [True for i in range(maxn + 1)]
        prime[0] = False
        prime[1] = False
        p = 2
        while (p * p <= maxn):
            if (prime[p] == True):
                for i in range(p * 2, maxn + 1, p):
                    prime[i] = False
            p += 1
            
        ohRight = []
        
        n = len(nums)
        
        for i in range(1, maxn):
            if prime[i]:
                ohRight.append(i)
                
        nums = [0] + nums
        
        ohRight = sorted(ohRight)[::-1]
        
        n += 1
        
        for i in range(1, n):
            for prime in ohRight:
                if nums[i] - prime > nums[i - 1]:
                    nums[i] = nums[i] - prime
                    break
                  
        for i in range(1, n):
            if nums[i] <= nums[i - 1]:
                return False
        
        return True

File: 2601-prime-subtraction-operation/2601-prime-subtraction-operation.cpp
class Solution {
public:
    bool primeSubOperation(vector<int>& nums) {
        const int MN = 1001;
        const int n = nums.size();
        vector<int> primes;
        bool visited[MN] = {false};
        visited[0] = visited[1] = true;
        
        for (int i = 2; i < MN; ++i) {
            if (!visited[i]) {
                for (int j = 2*i; j < MN; j += i) {
                    visited[j] = true; 
                }
                primes.push_back(i);
            }
        }
        
        int target = 0;
        
        for (int i = 0; i < n; ++i) {
            int p = 0;
            int left = 0, right = primes.size() - 1;
            
            while (left <= right) {
                int mid = (left + right) >> 1;
                if (nums[i] - primes[mid] > target) {
                    p = primes[mid];
                    left = mid + 1;
                }
                else {
                    right = mid - 1;
                }
            }
            
            if (nums[i] - p <= target) {
                return false;
            }
            
            target = nums[i] - p;
        }
        
        return true;
    }
};